"expected type, found constant " when using const generics with a `const` variable

A minimal reproduction is available here: https://github.com/arilotter/anchor-idl-bug

In the following code, `anchor build` throws 
```
error[E0573]: expected type, found constant `SOME_CONST`
   |
   |     pub checkpointers: Foo<I, SOME_CONST>,
   |                               ^^^^^^^^^^ not a type
```

but if you replace `SOME_CONST` with the literal `16`, there's no error.

```rs
use anchor_lang::prelude::*;

declare_id!("GETJtjFdo9LLt56QW681NPj3x74WuNhcFKULX2hU7eQc");

#[program]
pub mod idl_repro {
    use super::*;

    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
        ctx.accounts.coordinator.checkpointers = Foo::<16>::default();
        Ok(())
    }
}

pub const SOME_CONST: usize = 16;

#[account]
#[derive(
    Debug, Copy
)]
#[repr(C)]
pub struct Bar {
    // Using this const causes the error:
    // error[E0573]: expected type, found constant `SOME_CONST`
    //   --> programs/idl-repro/src/lib.rs:23:31
    //    |
    // 23 |     pub checkpointers: Foo<I, SOME_CONST>,
    //    |                               ^^^^^^^^^^ not a type
    //
    // However, if you replace SOME_CONST with the literal `16`, there's no error.
    pub checkpointers: Foo<SOME_CONST>, 
}

#[derive(Clone, Copy, Debug, AnchorDeserialize, AnchorSerialize,Default)]
pub struct Foo<const N: usize>;

#[derive(Accounts)]
pub struct Initialize<'info> {
    #[account(init, payer = user, space = 8 + std::mem::size_of::<Bar>())]
    pub coordinator: Account<'info, Bar>,
    #[account(mut)]
    pub user: Signer<'info>,
    pub system_program: Program<'info, System>,
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -113,6 +113,7 @@ The minor version will be incremented upon a breaking change and the patch versi
 - cli: Fix altering user-provided lib names ([#3467](https://github.com/coral-xyz/anchor/pull/3467)).
 - idl: Fix missing `program::seed` resolution ([#3474](https://github.com/coral-xyz/anchor/pull/3474)).
 - lang: Fix adding `derive`s and `repr`s to type alias definitions in `declare_program!` ([#3504](https://github.com/coral-xyz/anchor/pull/3504)).
+- idl: Fix using constant identifiers as generic arguments ([#3522](https://github.com/coral-xyz/anchor/pull/3522)).
 
 ### Breaking
 

--- a/lang/syn/src/idl/defined.rs
+++ b/lang/syn/src/idl/defined.rs
@@ -608,7 +608,7 @@ pub fn gen_idl_type(
                                 defined.extend(def);
                                 quote! { #idl::IdlGenericArg::Type { ty: #ty } }
                             }
-                            _ => return Err(anyhow!("Unsupported generic arugment: {arg:#?}")),
+                            _ => return Err(anyhow!("Unsupported generic argument: {arg:#?}")),
                         };
                         generics.push(generic);
                     }

--- a/lang/syn/src/idl/defined.rs
+++ b/lang/syn/src/idl/defined.rs
@@ -581,17 +581,36 @@ pub fn gen_idl_type(
             if let Some(segment) = path.path.segments.last() {
                 if let syn::PathArguments::AngleBracketed(args) = &segment.arguments {
                     for arg in &args.args {
-                        match arg {
+                        let generic = match arg {
+                            syn::GenericArgument::Const(c) => {
+                                quote! { #idl::IdlGenericArg::Const { value: #c.to_string() } }
+                            }
+                            // `MY_CONST` in `Foo<MY_CONST>` is parsed as `GenericArgument::Type`
+                            // instead of `GenericArgument::Const` because they're indistinguishable
+                            // syntactically, as mentioned in
+                            // https://github.com/dtolnay/syn/blob/bfa790b8e445dc67b7ab94d75adb1a92d6296c9a/src/path.rs#L113-L115
+                            //
+                            // As a workaround, we're manually checking to see if it *looks* like a
+                            // constant identifier to fix the issue mentioned in
+                            // https://github.com/coral-xyz/anchor/issues/3520
+                            syn::GenericArgument::Type(syn::Type::Path(p))
+                                if p.path
+                                    .segments
+                                    .last()
+                                    .map(|seg| seg.ident.to_string())
+                                    .map(|ident| ident.len() > 1 && ident == ident.to_uppercase())
+                                    .unwrap_or_default() =>
+                            {
+                                quote! { #idl::IdlGenericArg::Const { value: #p.to_string() } }
+                            }
                             syn::GenericArgument::Type(ty) => {
                                 let (ty, def) = gen_idl_type(ty, generic_params)?;
-                                generics.push(quote! { #idl::IdlGenericArg::Type { ty: #ty } });
                                 defined.extend(def);
+                                quote! { #idl::IdlGenericArg::Type { ty: #ty } }
                             }
-                            syn::GenericArgument::Const(c) => generics.push(
-                                quote! { #idl::IdlGenericArg::Const { value: #c.to_string() } },
-                            ),
-                            _ => (),
-                        }
+                            _ => return Err(anyhow!("Unsupported generic arugment: {arg:#?}")),
+                        };
+                        generics.push(generic);
                     }
                 }
             }

--- a/tests/idl/programs/new-idl/src/lib.rs
+++ b/tests/idl/programs/new-idl/src/lib.rs
@@ -369,10 +369,14 @@ pub struct SomeStruct {
     pub field: u16,
 }
 
+/// Compilation check for the issue described in https://github.com/coral-xyz/anchor/issues/3520
+// TODO: Use this from client-side (instead of hardcoding) once `program.constants` is supported
+const GENERIC_CONST: usize = 8;
+
 #[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
 pub struct GenericStruct<T, const N: usize> {
     arr: [T; N],
-    sub_field: SubGenericStruct<8, T, Vec<Option<T>>>,
+    sub_field: SubGenericStruct<GENERIC_CONST, T, Vec<Option<T>>>,
 }
 
 #[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]

